synapse/federation/federation_server.py

```diff
@@ -586,6 +588,95 @@ async def on_send_leave_request(self, origin: str, content: JsonDict) -> dict:
         await self.handler.on_send_leave_request(origin, pdu)
         return {}
 
+    async def on_make_knock_request(
+        self, origin: str, room_id: str, user_id: str, supported_versions: List[str]
+    ) -> Dict[str, Union[EventBase, str]]:
+        """We've received a /make_knock/ request, so we create a partial knock
+        event for the room and hand that back, along with the room version, to the knocking
+        homeserver. We do *not* persist or process this event until the other server has
+        signed it and sent it back.
+
+        Args:
+            origin: The (verified) server name of the requesting server.
+            room_id: The room to create the knock event in.
+            user_id: The user to create the knock for.
+            supported_versions: The room versions supported by the requesting server.
+
+        Returns:
+            The partial knock event.
+        """
+        origin_host, _ = parse_server_name(origin)
+        await self.check_server_matches_acl(origin_host, room_id)
+
+        room_version = await self.store.get_room_version_id(room_id)
+        if room_version not in supported_versions:
+            logger.warning(
+                "Room version %s not in %s", room_version, supported_versions
+            )
+            raise IncompatibleRoomVersionError(room_version=room_version)
+
+        if not KNOWN_ROOM_VERSIONS[room_version].msc2403_knocking:
+            raise SynapseError(
+                403,
+                "This room version does not support knocking",
+                errcode=Codes.FORBIDDEN,
+            )
```

Comment:
There's a helper for this:

```suggestion
        room_version = await self.store.get_room_version(room_id)
        if not room_version.msc2403_knocking:
            raise SynapseError(
                403,
                "This room version does not support knocking",
                errcode=Codes.FORBIDDEN,
            )
```

Reply:
Ah, that's useful. Though the error message in the helper function doesn't quite make sense for this situation:

https://github.com/matrix-org/synapse/blob/4b965c862dc66c0da5d3240add70e9b5f0aa720b/synapse/storage/databases/main/state.py#L74

Something we can generalise separately though.

Reply:
Actually, this is a different error code as well, so I'm not sure how helpful this helper function would be (`UnsupportedRoomVersionError` vs `IncompatibleRoomVersionError`).

Reply:
Well, we need to ensure that the:

1. Room version is known to us (`get_room_version` does this automatically).
2. The room version is in `supported_versions` (I shouldn't have removed this in the above, I didn't realized that `supported_versions` is what the *other* server gave us).
3. The room version supports knocking.

The code I suggested handles 1 & 3, the current code handles 2 & 3.

Reply:
Aha, I was also missing that `supported_versions` was a check on the remote server's supported versions - which is separate from our own check.

Checking all three makes sense for `make_knock`, checking just 1&3 makes sense for `send_knock` (as there is no `supported_versions` parameter).

The code should now reflect this.